### PR TITLE
Fix ug gc profile

### PR DESCRIPTION
**Pull request by TammushB**

### CODE REVIEW

File: src/ChrCopyNumber.cpp [hunk 1]
```diff
@@ -562,12 +588,14 @@ void ChrCopyNumber::recalculateRatio(double *a, int degree) {
                 //use a threshold, but correct using notN profile
                 if (mappabilityProfile_.size()>0) {
                     if ((x>0)&&(mappabilityProfile_[i]>minMappabilityPerWindow)) //if ((x>0)&&(notNprofile_[i]!=0))
-                        ratio_[i] = ratio_[i]/polynomial(x,a,1,degree);
+                        //ratio_[i] = ratio_[i]/polynomial(x,a,1,degree);
+						ratio_[i] = ratio_[i]/1;
```
Comment (TammushB) [inline src/ChrCopyNumber.cpp]: @ilyasoifer this is the GC normalization disable. not sure if this is the best implementation. I can try adding a flag I guess...

File: src/ChrCopyNumber.cpp [hunk 2]
```diff
@@ -658,15 +689,27 @@ void ChrCopyNumber::recalculateLogRatio (float constant) {
...
 void ChrCopyNumber::recalculateRatioWithContam (float contamination, float normGenytype, bool isLogged) { //normGenytype==1 if AB, normGenytype==0.5 if A
-	if (!isLogged) {
-        for (int i = 0; i<length_; i++)
-            if (ratio_[i] != NA) {
-			//ratio_[i] = (ratio_[i]-contamination*normGenytype)/(1-contamination); //correct only for ploidy 2
-                ratio_[i] = (ratio_[i]*(1-contamination+2*contamination/ploidy_) -contamination*normGenytype/ploidy_*2)/(1-contamination);
-
-			if (ratio_[i]<0)
-				ratio_[i] = 0;
+	if (!isLogged) {
```
Comment (TammushB) [inline src/ChrCopyNumber.cpp:676]: @ilyasoifer this block changes include mainly prints of ug ratio correction. I think we can keep this for further debugging. it does not affect the run itself.

**Comment by TammushB** [comment]
@ilyasoifer  please review. 
main changes are : 
1. GC normalization disable. 
2. docker file + GH action to build it
3. script that generates the FREEC config file
4. NaiveNormalization mode 
5. Segments printing to a final output file. 

there are some debugging prints from within the pipeline - I think we can leave them  for now. 
all changes are for WGS, no log scale cases.

File: src/ChrCopyNumber.cpp [hunk 3]
```diff
@@ -467,6 +467,32 @@ void ChrCopyNumber::calculateRatio(ChrCopyNumber control, const double * a, cons
 
 			if (mappabilityProfile_.size() == 0 || mappabilityProfile_[i] > minMappabilityPerWindow) {
                 ratio_[i] = readCount_[i]/polynomial(control.getValueAt(i),a,1,degree);
```
Comment (ilyasoifer) [inline src/ChrCopyNumber.cpp]: Something meesed up in the indentation

File: src/freec
Comment (ilyasoifer) [inline src/freec:1]: Was this compiled file added by mistake?

File: src/Makefile.freec [hunk 4]
```diff
@@ -6,7 +6,7 @@
 # Eric Viara, Valentina Boeva January 2013
 #
 
-CXXOPT = -O3 -g
```
Comment (ilyasoifer) [inline src/Makefile.freec:9]: Do you want to revert -O3?

File: src/main.cpp [hunk 5]
```diff
@@ -267,6 +267,10 @@ int main(int argc, char *argv[])
 
     bool has_window = cf.hasValue("general","window");
     int window = (int)cf.Value("general","window",NA);
+    bool naiveNormalization = cf.Value("general", "NaiveNormalization",false);
```
Comment (ilyasoifer) [inline src/main.cpp:270]: Perhaps add this to the README?

File: src/main.cpp [hunk 6]
```diff
@@ -1190,7 +1194,9 @@ int main(int argc, char *argv[])
             }
         }
 
-	sampleCopyNumber.printRatio(myName+"_ratio.txt",0,printNA);
+	
+    sampleCopyNumber.printSegments(myName+"_segments.txt");
```
Comment (ilyasoifer) [inline src/main.cpp:1198]: Indentation

File: src/GenomeCopyNumber.cpp [hunk 7]
```diff
@@ -1171,20 +1174,29 @@ long double GenomeCopyNumber::calculateRSS(int ploidy)
...
+            MyFile << observedvalues[i] << "\t" << expectedvalues[i] << "\t" << expected_round << "\t" << diff << "\t" << RSS_value << "\n";
             }
         }
+    MyFile.close();
+
     if (observedvalues.size()==0) {
         return 0;
     }
     double normRSS = (RSS/observedvalues.size());
+    cout << "normRSS = " << normRSS << "\n";
```
Comment (ilyasoifer) [inline src/GenomeCopyNumber.cpp]: Debugging print?

File: src/GenomeCopyNumber.cpp [hunk 8]
```diff
@@ -1171,20 +1174,29 @@ long double GenomeCopyNumber::calculateRSS(int ploidy)
 	}
 
     long double RSS = 0;
+    auto out_file_name = "RSS_expected_observed.p" + std::to_string(ploidy) + ".tsv";
```
Comment (ilyasoifer) [inline src/GenomeCopyNumber.cpp]: Is this really needed or only for debugging?

File: src/GenomeCopyNumber.cpp [hunk 9]
```diff
@@ -1309,9 +1321,61 @@ int GenomeCopyNumber::fillInRatio() {
...
+        vector <float> y_unalternated;
+        for ( it=chrCopyNumber_.begin() ; it != chrCopyNumber_.end(); it++ ){
+            if (! (it->getChromosome().find("X")!=string::npos || it->getChromosome().find("Y")!=string::npos)){
+                vector <float> controlcounts = controlCopyNumber.getChrCopyNumber(it->getChromosome()).getValues() ;
+                for (int i = 0; i< it->getLength(); i++) {
+                    if (it->getValueAt(i)>0) {
+                        float rc_sample = it->getValueAt(i);
+                        // float naive_ratio = (rc_sample/sample_median)/(controlcounts[i]/control_median);
+                        float naive_ratio = (rc_sample/sample_mean)/(controlcounts[i]/control_mean);
+                        if((naive_ratio>0.97) && (naive_ratio<1.03)){
```
Comment (ilyasoifer) [inline src/GenomeCopyNumber.cpp]: Hardcoded thresholds... At least make them constants

File: src/GenomeCopyNumber.cpp [hunk 10]
```diff
@@ -1309,9 +1321,61 @@ int GenomeCopyNumber::fillInRatio() {
...
+                }
+                controlcounts.clear();
+            }
+        }
+        sample_median = get_median(y_unalternated);
+        sample_mean = get_mean(y_unalternated);
+        cout << "sample median value: " << sample_median << "\n";
 
+        for ( it=chrCopyNumber_.begin() ; it != chrCopyNumber_.end(); it++ ) {
+                // it->calculateRatio(controlCopyNumber.getChrCopyNumber(it->getChromosome()),control_median,sample_median);
```
Comment (ilyasoifer) [inline src/GenomeCopyNumber.cpp]: leftover

File: src/GenomeCopyNumber.cpp [hunk 11]
```diff
@@ -1309,9 +1321,61 @@ int GenomeCopyNumber::fillInRatio() {
...
+                for (int i = 0; i< it->getLength(); i++) {
+                    if (it->getValueAt(i)>0) {
+                        x.push_back(controlcounts[i]);
+                        y.push_back(it->getValueAt(i));
+                    }
+                }
+                controlcounts.clear();
+            }
+        }
+        float control_median = get_median(x);
```
Comment (ilyasoifer) [inline src/GenomeCopyNumber.cpp]: Is median used?

File: src/GenomeCopyNumber.cpp [hunk 12]
```diff
@@ -2983,6 +3048,44 @@ void GenomeCopyNumber::calculateSomaticCNVs (std::vector <EntryCNV> controlCNVs,
...
+void GenomeCopyNumber::printSegments(std::string const& outFile) {
+    std::ofstream file (outFile.c_str());
+    map<string,int>::iterator it;
+
+	file << "chr\tstart\tend\tmedian_ratio\tsd_ratio\n";
+    for (int i = 0; i < (int)chrCopyNumber_.size(); i++) {
+        printSegments(chrCopyNumber_[i],file);
+	}
+	file.close();
+    // // for ( it=chrCopyNumber_.begin() ; it != chrCopyNumber_.end(); it++ ) {
```
Comment (ilyasoifer) [inline src/GenomeCopyNumber.cpp]: Leftovers?

File: src/ChrCopyNumber.cpp [hunk 13]
```diff
@@ -467,6 +467,32 @@ void ChrCopyNumber::calculateRatio(ChrCopyNumber control, const double * a, cons
...
+			} else
+                ratio_[i] = NA;
+
+		} else
+			if (readCount_[i]==0)
+				//ratio_[i] = 1;
+				ratio_[i] = NA;
+			else
+				ratio_[i] = NA;
+		//cout << readCount_[i] << "\t" << control.getValueAt(i) << "\t" << ratio_[i] << "\n";
```
Comment (ilyasoifer) [inline src/ChrCopyNumber.cpp]: Debug print leftover

File: src/ChrCopyNumber.cpp [hunk 14]
```diff
@@ -467,6 +467,32 @@ void ChrCopyNumber::calculateRatio(ChrCopyNumber control, const double * a, cons
...
                 ratio_[i] = readCount_[i]/polynomial(control.getValueAt(i),a,1,degree);
+				
+				if (ratio_[i]<0)
+					ratio_[i] = NA;
+			} else
+                ratio_[i] = NA;
+
+		} else
+			if (readCount_[i]==0)
+				//ratio_[i] = 1;
```
Comment (ilyasoifer) [inline src/ChrCopyNumber.cpp]: Leftover

File: src/ChrCopyNumber.cpp [hunk 15]
```diff
@@ -562,12 +588,14 @@ void ChrCopyNumber::recalculateRatio(double *a, int degree) {
                 //use a threshold, but correct using notN profile
                 if (mappabilityProfile_.size()>0) {
```
Comment (ilyasoifer) [inline src/ChrCopyNumber.cpp]: Can you fix the indentation here please?

File: src/ChrCopyNumber.cpp [hunk 16]
```diff
@@ -467,6 +467,32 @@ void ChrCopyNumber::calculateRatio(ChrCopyNumber control, const double * a, cons
...
+
+
+void ChrCopyNumber::calculateRatio(ChrCopyNumber control,const float control_median,const float sample_median){
+	if ((int)ratio_.size()!=length_)
+		ratio_.resize(length_);
+	for (int i = 0; i<length_; i++) {
+		if ((control.getLength()>i)&&(control.getValueAt(i) != 0)){
+
+			if (mappabilityProfile_.size() == 0 || mappabilityProfile_[i] > minMappabilityPerWindow) {
+				ratio_[i] = (readCount_[i]/sample_median)/(control.getValueAt(i)/control_median);
```
Comment (ilyasoifer) [inline src/ChrCopyNumber.cpp:488]: Why is here ratio_ being normalized in a different way (not through a polynomial)?

File: src/ChrCopyNumber.cpp (hunk 1 above)
Comment (ilyasoifer) [reply src/ChrCopyNumber.cpp]: I think it could be cleaner to fix this at the step that produces polynomial (make it a constant 1)

File: src/ChrCopyNumber.cpp [hunk 17]
```diff
@@ -593,20 +621,23 @@ void ChrCopyNumber::calculateRatio(double *a, int degree) {
 			if (uniqueMatch) {
                 //use mappabilityProfile_ and correct
                 if ((x>0)&&(mappabilityProfile_[i]>minMappabilityPerWindow)) //if ((x>0)&&(notNprofile_[i]!=0))
-                    ratio_[i] = readCount_[i]/polynomial(x,a,1,degree)/mappabilityProfile_[i];
+                    //ratio_[i] = readCount_[i]/polynomial(x,a,1,degree)/mappabilityProfile_[i];
+					ratio_[i] = readCount_[i]/1/mappabilityProfile_[i];
```
Comment (ilyasoifer) [inline src/ChrCopyNumber.cpp]: Same here about the polynomial

File: src/ChrCopyNumber.cpp [hunk 18]
```diff
@@ -593,20 +621,23 @@ void ChrCopyNumber::calculateRatio(double *a, int degree) {
...
                 else
                     ratio_[i] = NA;
 
 			} else {
                 //use a threshold, but correct using notN profile
                 if (mappabilityProfile_.size()>0) {
                     if ((x>0)&&(mappabilityProfile_[i]>minMappabilityPerWindow)) //if ((x>0)&&(notNprofile_[i]!=0))
-                        ratio_[i] = readCount_[i]/polynomial(x,a,1,degree)/notNprofile_[i];
+                        //ratio_[i] = readCount_[i]/polynomial(x,a,1,degree)/notNprofile_[i];
```
Comment (ilyasoifer) [inline src/ChrCopyNumber.cpp]: Same here about the polynomial

File: src/ChrCopyNumber.cpp (hunk 2 above)
Comment (ilyasoifer) [reply src/ChrCopyNumber.cpp:676]: OK, sure, we can keep this for now

File: src/ChrCopyNumber.cpp [hunk 19]
```diff
@@ -709,6 +752,23 @@ int ChrCopyNumber::getCoveredPart(int breakPointStart, int breakPointEnd) { //fo
 }
 
 void ChrCopyNumber::calculateCopyNumberMedian(int ploidy, int minCNAlength, bool noisyData,  bool CompleteGenomicsData, bool isLogged){ //create median profiles using 'bpfinal_' and store them in medianProfile_, info about medians themselves is stored in medianValues_ and about SD in sd_, lengths of fragments in bpLengths_
+	
+	auto out_file_name = "segments_info.p" + std::to_string(ploidy) + ".tsv";
+	ofstream MyFile(out_file_name,std::ios::app);
```
Comment (ilyasoifer) [inline src/ChrCopyNumber.cpp]: Can this be myFile?

File: src/ChrCopyNumber.cpp [hunk 20]
```diff
@@ -709,6 +752,23 @@ int ChrCopyNumber::getCoveredPart(int breakPointStart, int breakPointEnd) { //fo
...
+	
+	// std::ifstream MyFile(out_file_name);
+	// // ofstream MyFile;
```
Comment (ilyasoifer) [inline src/ChrCopyNumber.cpp]: Debug prints

File: src/ChrCopyNumber.h [hunk 21]
```diff
@@ -35,6 +35,7 @@ class ChrCopyNumber
     void recalculateRatio(ChrCopyNumber control);
 	void calculateRatio(ChrCopyNumber control, double a0, double a1);
 	void calculateRatio(ChrCopyNumber control, const double * a, const int degree);
+	void calculateRatio(ChrCopyNumber control,const float control_median,const float sample_median);
```
Comment (ilyasoifer) [inline src/ChrCopyNumber.h:38]: Can you say why this change was needed?

File: src/GenomeCopyNumber.cpp [hunk 22]
```diff
@@ -3558,35 +3661,35 @@ void GenomeCopyNumber::addBAFinfo(SNPinGenome & snpingenome) {
...
-//                if (SNPcount < totalSNPnumber) {
-//                    SNPcount++;
-//                    getSNPpos = snpingenome.SNP_atChr(indexSNP).getPositionAt(SNPcount);
-//                    currentBAF = snpingenome.SNP_atChr(indexSNP).getValueAt(SNPcount);
-//                    i=max(-1,i-windowSize_/step_-1);
-//                    //cout << SNPcount << " out of "<< totalSNPnumber<<"\n";
-//                }
-//            }
-//		}
+		chrCopyNumber_[index].createBAF(NA);
```
Comment (ilyasoifer) [inline src/GenomeCopyNumber.cpp:3654]: Can you explain this change?

File: src/Makefile.freec [hunk 23]
```diff
@@ -6,7 +6,7 @@
...
+CXXOPT = -g
```
Comment (ilyasoifer) [inline src/Makefile.freec]:
```suggestion
CXXOPT = -g -O3 
```

File: src/myFunc.cpp [hunk 24]
```diff
@@ -1552,11 +1552,29 @@ float runEM (const vector<float>& x,const vector<float>& y,double * a, int degre
 
 	for (count = 0; count < NumberOfIterations; count++) {
 
+		// assign best ploidy for each window 
+		std::vector<float> res_at_ploidy;
```
Comment (ilyasoifer) [inline src/myFunc.cpp:1556]: Are we still using this change?

**Review by ilyasoifer — COMMENTED** [review]
Hi, impressive amount of work. Can you take a look and see around my comments if all changes are still needed? 
I am worried about submitting this back to ControlFREEC

Also - could you update their README please?

File: src/freec
Comment (TammushB) [reply src/freec:1]: Actually this is not a mistake. 
we need to add the compiled files for the github action that builds the FREEC docker.

File: src/Makefile.freec (hunk 4 above)
Comment (TammushB) [reply src/Makefile.freec:9]: I don't see why. it really improved debugging  and it is not harming the executable optimization.
Claude says: "This configuration is often used for release builds, where you want to have a highly optimized executable that still retains the ability to be debugged if necessary."

File: src/main.cpp (hunk 5 above)
Comment (TammushB) [reply src/main.cpp:270]: there is no parameters documentation in the README file. 
added this parameter WGS config example. 
also added a notation for the segments output in the README.

File: src/GenomeCopyNumber.cpp (hunk 8 above)
Comment (TammushB) [reply src/GenomeCopyNumber.cpp]: this print shows us the effect of correction for ploidy for each ploidy value. 
if we aim at trying ploidy estimation again this can be useful. 
removing for now.

File: src/ChrCopyNumber.cpp (hunk 16 above)
Comment (TammushB) [reply src/ChrCopyNumber.cpp:488]: this is my addition for naive normalization. 
in case NaiveNormalization=TRUE , the pipeline will use this usecase of calculateRatio function.

File: src/ChrCopyNumber.cpp (hunk 17 above)
Comment (TammushB) [reply src/ChrCopyNumber.cpp]: polynomial is used for normalization of read counts and for normalization of GC profile. 
I am worried that changing the value from within _polynomial_ function will affect the values of coverage normalization. 
We made this change because we wanted to avoid GCprofile normalization. yet, the pipeline kind of can't run properly without it (bug? or I missed something). 
What do you think of adding a flag : UG_GC_Normalization into the conf? and use it within the code to generate pseudo GC profile ?

File: src/ChrCopyNumber.h (hunk 21 above)
Comment (TammushB) [reply src/ChrCopyNumber.h:38]: we created a new calculateRatio function that with this set of inputs calculates naive fold-change ((sample/mean)/(control/mean)). 
this function will be used when NaiveNormalization=TRUE in the input config file.

File: src/myFunc.cpp (hunk 24 above)
Comment (TammushB) [reply src/myFunc.cpp:1556]: yes, this is used for  the case of choosing with running with polynomial fit. 
I think this is a reasonable contribution.

File: src/ChrCopyNumber.cpp (hunk 1 above)
Comment (TammushB) [reply src/ChrCopyNumber.cpp]: see my comment about using polynomial for GCprofile

File: src/GenomeCopyNumber.cpp (hunk 22 above)
Comment (TammushB) [reply src/GenomeCopyNumber.cpp:3654]: this is strange. I don't remember touching this block...

**Comment by TammushB** [comment]
@ilyasoifer thanks for the comments. 
I think that the most problematic issue to contribute back is the GC normalization issue. 
and maybe the fact that all changed are for WGS sample-control use case.

File: src/GenomeCopyNumber.cpp (hunk 22 above)
Comment (TammushB) [reply src/GenomeCopyNumber.cpp:3654]: I think I removed the comment in order to generate output BAF file.

File: src/ChrCopyNumber.cpp (hunk 1 above)
Comment (TammushB) [reply src/ChrCopyNumber.cpp]: looking at this again , I see that in current flow (both regular and high sensitivity mode) we are not stepping into this function. reverting back to polynomial of GCprofile.

**Comment by TammushB** [comment]
merging this. 
notes for further merging with FREEC master : 
- our changes apply for WGS not log data usecase. 
- remove debugging prints.
